jobs: replace the ie in Registry.createJobsInBatchWithTxn

ZhouXing19 opened the pull request:
When creating a job, we should use an internal executor bound to the outer txn and related metadata.

We also found that in this case, if a user changes the timezone and then triggers a schema change job, the job's creation time (i.e.   value `created`column in the `system.jobs`) will automatically switch to the timestamp in the updated timezone, rather than UTC. This is because we have the `created` using `now()` in the SQL statement. We now change it to be inserted always with UTC time.

Informs: #91004
Informs: #91718

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91679)
<!-- Reviewable:end -->

ZhouXing19 committed:
jobs: replace the ie in Registry.createJobsInBatchWithTxn

When creating a job, we should use an internal executor bound to the outer txn
and related metadata.

We also found that in this case, if a user changes the timezone and then triggers
a schema change job, the job's creation time (i.e. value createdcolumn in the
system.jobs) will automatically switch to the timestamp in the updated timezone,
rather than UTC. This is because we have the created using now() in the SQL
statement. We now change it to be inserted always with UTC time.

Informs: #91004
Informs: #91718

Release note: None

Xiang-Gu commented on pkg/jobs/adopt.go:
```diff
@@ -72,19 +75,22 @@ func (r *Registry) maybeDumpTrace(
 	// that has failed from a job that has been canceled.
 	if jobErr != nil && !HasErrJobCanceled(jobErr) && resumerCtx.Err() == nil {
-		r.td.Dump(dumpCtx, strconv.Itoa(int(jobID)), traceID, r.ex)
+
```
nit: extra empty line

Xiang-Gu commented on pkg/jobs/adopt.go:
```diff
@@ -124,6 +130,16 @@ const (
 	canRunArgs = `(SELECT $3::TIMESTAMP AS ts, $4::FLOAT AS initial_delay, $5::FLOAT AS max_delay) args`
 	// NextRunClause calculates the next execution time of a job with exponential backoff delay, calculated
 	// using last_run and num_runs values.
```
nit: This comment belongs to the query below I believe.

nit: Add comment for this newly added query?

ajwerner reviewed (commented):
Instead of this approach, can we leave the query exactly as it is and change the way the value which gets generated comes to be by not using the default value? Consider:

```go
diff --git a/pkg/jobs/registry.go b/pkg/jobs/registry.go
index 24a96699ad1..55b7dfed541 100644
--- a/pkg/jobs/registry.go
+++ b/pkg/jobs/registry.go
@@ -379,7 +379,7 @@ func (r *Registry) batchJobInsertStmt(
 ) (string, []interface{}, []jobspb.JobID, error) {
 	instanceID := r.ID()
 	const numColumns = 6
-	columns := [numColumns]string{`id`, `status`, `payload`, `progress`, `claim_session_id`, `claim_instance_id`}
+	columns := [numColumns]string{`id`, `created`, `status`, `payload`, `progress`, `claim_session_id`, `claim_instance_id`}
 	marshalPanic := func(m protoutil.Message) []byte {
 		data, err := protoutil.Marshal(m)
 		if err != nil {
@@ -387,8 +387,13 @@ func (r *Registry) batchJobInsertStmt(
 		}
 		return data
 	}
+	created, err := tree.MakeDTimestamp(timeutil.FromUnixMicros(modifiedMicros), time.Microsecond)
+	if err != nil {
+		return "", nil, nil, errors.NewAssertionErrorWithWrappedErrf(err, "failed to make timestamp")
+	}
 	valueFns := map[string]func(*Record) interface{}{
 		`id`:                func(rec *Record) interface{} { return rec.JobID },
+		`created`:           func(record *Record) interface{} { return created },
 		`status`:            func(rec *Record) interface{} { return StatusRunning },
 		`claim_session_id`:  func(rec *Record) interface{} { return sessionID.UnsafeBytes() },
 		`claim_instance_id`: func(rec *Record) interface{} { return instanceID },
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-:-NGc8mtn1BY1Pu3Nut8g:b-gvranc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rhu713 and @ZhouXing19)

___
*[`pkg/jobs/adopt.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-NGc8iLX1wEH4sb1MJAM:-NGc8iLX1wEH4sb1MJAN:b-d8fkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/52016819006530e7bf1217a5c462ee57af460fa4/pkg/jobs/adopt.go#L135)):*
> ```Go
> 	PossibleDelay = `args.initial_delay * (power(2, least(62, COALESCE(num_runs, 0))) - 1)::FLOAT`
> 	Piece1        = `COALESCE(last_run, created)`
> 	Piece2        = `least(
> ```

nit: don't export these



_Code quote:_
```Go
	PossibleDelay = `args.initial_delay * (power(2, least(62, COALESCE(num_runs, 0))) - 1)::FLOAT`
	Piece1        = `COALESCE(last_run, created)`
	Piece2        = `least(
```


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> change the way the value which gets generated comes to be by not using the default value

Oh I did something similar in the second commit already: https://github.com/cockroachdb/cockroach/pull/91679/commits/82f2c6ecc3cf81eeba3ef102963e1917aafa4500#
But I was with `r.clock.Now().GoTime().Format(time.RFC3339Nano)`. Do you think that would work too?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-:-NGcHLh97m4mndAxrC7-:bmea8pc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rhu713)



<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> can we leave the query exactly as it is

Removed the first commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-:-NGcJQCo02WptJTbxHle:byd77ig)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rhu713 and @Xiang-Gu)

___
*[`pkg/jobs/adopt.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-NGXILVe3T7ttRbXfp7o-r1-85:-NGcJVRM7rEOLhJk5fEa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52016819006530e7bf1217a5c462ee57af460fa4/pkg/jobs/adopt.go#L85)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

nit: extra empty line
</blockquote></details>

Done.
___
*[`pkg/jobs/adopt.go` line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-NGXILVe3T7ttRbXfp7o-r1-132:-NGcJWDG6hnrswZfxx13:by6yp8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/52016819006530e7bf1217a5c462ee57af460fa4/pkg/jobs/adopt.go#L132)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

nit: This comment belongs to the query below I believe.

nit: Add comment for this newly added query?
</blockquote></details>

Those are queries that I used for debugging and forgot to remove them in that commit :sweat_smile: Have them deleted now.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod the nit

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-:-NGcKocbAovgOd1WmCh3:brupp8d)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rhu713, @Xiang-Gu, and @ZhouXing19)

___
*[`pkg/jobs/registry.go` line 397 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-NGcLNjb53vRU_mXekWP:-NGcLNjb53vRU_mXekWQ:b-5iinai) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4311181db9fef87a7d22b0d10d41a0cc67403f/pkg/jobs/registry.go#L397)):*
> ```Go
> 	valueFns := map[string]func(*Record) interface{}{
> 		`id`:                func(rec *Record) interface{} { return rec.JobID },
> 		`created`:           func(rec *Record) interface{} { return r.clock.Now().GoTime().Format(time.RFC3339Nano) },
> ```

I think this should use `modifiedMicros`. It'll evaluate to the same value as `now()` which it relative to the transaction timestamp. I also have a mild preference for constructing the datum as opposed to passing the string literal to the sql layer and making it parse it back. It also alleviates the need to ask the question whether `RFC3339Nano` is a format that sql likes. 


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
*[`pkg/jobs/registry.go` line 397 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91679#-NGcLNjb53vRU_mXekWP:-NGcPyR5DqoD7N61jnh3:b-l6uj7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4311181db9fef87a7d22b0d10d41a0cc67403f/pkg/jobs/registry.go#L397)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think this should use `modifiedMicros`. It'll evaluate to the same value as `now()` which it relative to the transaction timestamp. I also have a mild preference for constructing the datum as opposed to passing the string literal to the sql layer and making it parse it back. It also alleviates the need to ask the question whether `RFC3339Nano` is a format that sql likes. 
</blockquote></details>

That makes sense, thanks for explaining! 
Done.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
TFTR!
bors r=ajwerner
